text/0000-hash/0000-hash.md

```diff
@@ -0,0 +1,68 @@
++ Feature name: `bee-hash`
++ Start date: 2019-10-15
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)
+
+# Summary
+
+Hash functions in IOTA are sponge construction which usually hash an internal start to absorb input stream and output
+stream in any desired length after permutation. That means they all share similar structure we are able to abstract
+with. The goal of this RFC is to model this Sponge layer as trait. Making low-level hash algorithms provide same logic
+interface exposed to user, and let high-level module able to loosely couple from them.
+
+# Motivation
+
+It's true that everyone can just use each hash function they want directly. But we should still prevent from any of
+these algorithm is too tightly coupled to upper module when we integrate this. Take proof of work for example, it will
+have a module called PearlDiver mainly to deal with hashing. In the past, it already switched from curl-p to kerl.
+Imagine these two functions have totally different API, it would be difficult to refactor. Same thing is going to happen
+when we migrate to troika or other secure cryptography primitives in the future. So defining a trait for sponge is the
+best and necessary approach for us. 
+
+# Detailed design
+
+This trait is going to be the interface of any sponge construction. Imagine it's like a real sponge, it `absorb`s the
+clone of input reference provided by users. And they don't need to know how internal permutation and transformation
+work, just `squeeze` the sponge and there will be a mutable reference as output. The internal state will not be cleared
+unless `reset` is called.
+
+```rust 
+pub trait Sponge
+where
+    Self: Default + Clone + Send + 'static,
```

Comment:
Are these bounds (`Send + 'static`) required for *all* sponges? If a function requires a `Send`able sponge, it can just accept `impl Sponge + Send + 'static`.

Reply:
Thanks! I will just leave `Default + Clone` here. I think these are required at least.